sharding/xorValidatorShuffler.go

```diff
@@ -102,20 +102,28 @@ func (rxs *randXORShuffler) UpdateNodeLists(args ArgsUpdateNodes) (map[uint32][]
 }
 
 func removeFromWaiting(waitingNodes map[uint32][]Validator, leavingNodes []Validator) (map[uint32][]Validator, []Validator) {
-	for shard, vList := range waitingNodes {
-		nbToRemove := len(vList)
-		if len(leavingNodes) < nbToRemove {
-			nbToRemove = len(leavingNodes)
-		}
-
-		vList, removedNodes := removeValidatorsFromList(vList, leavingNodes, nbToRemove)
-		leavingNodes, _ = removeValidatorsFromList(leavingNodes, removedNodes, len(removedNodes))
-		waitingNodes[shard] = vList
+	for shard := uint32(0); shard < uint32(len(waitingNodes)-1); shard++ {
+		leavingNodes = computeLeaving(waitingNodes, leavingNodes, shard)
 	}
 
+	leavingNodes = computeLeaving(waitingNodes, leavingNodes, core.MetachainShardId)
+
 	return waitingNodes, leavingNodes
 }
 
+func computeLeaving(waitingNodes map[uint32][]Validator, leavingNodes []Validator, shard uint32) []Validator {
```

Comment:
this doesn't just compute, it also updates the waiting nodes

Reply:
renamed to removeNodes